pkg/bff/mock/admin.go

```diff
@@ -0,0 +1,162 @@
+	// Configure the router for the mock
+	gin.SetMode(gin.TestMode)
+	a.router = gin.New()
+	a.authorize = apiv2.Authorization(a.tokens)
+	a.csrf = apiv2.DoubleCookie()
+	a.setupHandlers()
+
+	a.srv = httptest.NewTLSServer(a.router)
+	if a.client, err = admin.NewMock(a.tokens, a.srv); err != nil {
+		return nil, err
+	}
+
+	return a, nil
+}
+
+// Admin is a mock implementation the DirectoryAdministrationServer using an httptest
+// server. The Client() method provides access to the DirectoryAdministrationClient
+// that can be used by tests to make calls to the Admin API using the BFF client. This
+// allows the BFF tests to set specific responses for individual endpoints. To
+// configure the behavior for specific endpoints, use the UseFixture, UseError, or
+// UseHandler methods.
+type Admin struct {
+	srv       *httptest.Server
+	router    *gin.Engine
+	client    apiv2.DirectoryAdministrationClient
+	tokens    *tokens.TokenManager
+	authorize gin.HandlerFunc
+	csrf      gin.HandlerFunc
+	handlers  map[string]gin.HandlerFunc
+	Calls     map[string]int
+}
+
+// UseFixture allows tests to use a JSON fixture from disk as the response.
+func (a *Admin) UseFixture(endpoint, path string) (err error) {
+	// Read the fixture data from disk
+	var data []byte
+	if data, err = ioutil.ReadFile(path); err != nil {
+		return fmt.Errorf("could not read fixture data: %s", err)
+	}
+
+	switch endpoint {
+	case StatusEP:
+		out := &apiv2.StatusReply{}
+		if err = json.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal fixture data: %s", err)
+		}
+		if err = a.handle(endpoint, func(c *gin.Context) {
+			c.JSON(http.StatusOK, out)
+		}); err != nil {
+			return fmt.Errorf("could not handle fixture data: %s", err)
+		}
+	case ListCertificatesEP:
+		out := &apiv2.ListCertificatesReply{}
+		if err = json.Unmarshal(data, out); err != nil {
+			return fmt.Errorf("could not unmarshal fixture data: %s", err)
+		}
+		if err = a.handle(endpoint, func(c *gin.Context) {
+			c.JSON(http.StatusOK, out)
+		}); err != nil {
+			return fmt.Errorf("could not set handler: %s", err)
+		}
+	default:
+		return fmt.Errorf("unsupported endpoint: %s", endpoint)
+	}
+
+	return nil
+}
+
+// UseError allows tests to specify an HTTP error for the response.
+func (a *Admin) UseError(endpoint string, code int, msg string) (err error) {
+	return a.handle(endpoint, func(c *gin.Context) {
+		c.JSON(code, msg)
+	})
+}
+
+// UseHandler allows tests to specify a custom handler for the endpoint.
+func (a *Admin) UseHandler(endpoint string, handler gin.HandlerFunc) (err error) {
+	return a.handle(endpoint, handler)
+}
+
+// handle is a helper function that adds a handler to the handlers map and
+// returns that handler function when the endpoint is called.
+func (a *Admin) handle(endpoint string, handler gin.HandlerFunc) error {
+	if _, ok := a.handlers[endpoint]; !ok {
+		return fmt.Errorf("unhandled endpoint %s", endpoint)
+	}
+	a.handlers[endpoint] = handler
+	return nil
```

Comment:
The error handling is a bit unwieldy here, may I just suggest a panic instead? The audience for this code is a test user and if they handle an endpoint that's not in the mock, a panic makes sense. In the gRPC mocks, rather than using a handlers map, we use methods like `OnStatus`, and `OnListCertificates` that are set to `nil` and cause a panic if called by the client; so the panic is consistent with our other test mechanisms.

Reply:
Sure, that makes sense to me.